sunpy/net/dataretriever/client.py

```diff
@@ -137,86 +101,67 @@ class GenericClient(BaseClient):
     `~sunpy.net.dataretriever.client.GenericClient.fetch` the former generates a
     set of results for files available through the service the client is
     querying and the latter downloads that data.
-
-    The `~sunpy.net.dataretriever.client.GenericClient.search` method takes a
-    set of `sunpy.net.attrs` objects and then converts these into a call to
-    `~sunpy.net.dataretriever.client.GenericClient._get_url_for_timerange`. It
-    does this through the `map\\_` dictionary which represents the
-    `~sunpy.net.attrs` objects as a dictionary.
     """
+    baseurl = None
+    pattern = None
+    required = {a.Time, a.Instrument}
+    optional = set()
 
-    def __init__(self):
-        self.map_ = {}
-
-    def _makeargs(self, *args):
+    @classmethod
+    def _get_match_dict(cls, *args, **kwargs):
         """
-        Construct the `map\\_` internal representation of the query.
-
-        This `map\\_` dictionary is passed through to the
-        `_get_url_for_timerange` method to get the URL results.
+        Constructs a dictionary using the query and registered Attrs that represents
+        all possible values of the extracted metadata for files that matches the query.
+        The returned dictionary is used to validate the metadata of searched files
+        in `:func:~sunpy.util.scraper.Scraper._extract_files_meta`.
 
         Parameters
         ----------
         \\*args: `tuple`
-            The query attributes.
+            `sunpy.net.attrs` objects representing the query.
+        \\*\\*kwargs: `dict`
+             Any extra keywords to refine the search.
 
+        Returns
+        -------
+        matchdict: `dict`
+            A dictionary having a `list` of all possible Attr values
```

Comment:
It knows them all by `register_values()`. Only those which are registered in that method will be known to the client.

Reply:
This seems flawed. The big change to GOES in particular was us being able to search all the available satellites, but because we are only search for ones which we have listed in `register_values` if a new satellite becomes operational we would need the user to upgrade sunpy before they could get results for that satellite.

This pulls us back to redesigning scraper doesn't it?

Reply:
I think it's probably ok given that its somewhat of a rare occurrence that there'll be a new GOES satellite with science quality XRS data so I think this is ok to have it 'hard-coded' into the register_values. 

Like the more I think about this - people really will only want just GOES data from one instrument given that they are _supposed_ to be the same data (calibrated to be the same anyway). Those that are interested in using other GOES satellites (e.g. wanting to use GOES 13 when there is GOES 15) will could just use the `SatelliteNumber` attrs. The majority of the community will really just want the lightcurves from the most current instrument at the time of their observation. We could just have a nice example of how to do this etc. 

But as @abhijeetmanhas mentioned in the matrix chat - how would scraper work when the pattern is `../satellite_number/y/m/d...` which is the case for GOES 16 and 17 (rather than just listing all files under `../goes_data/fits/y/m/d..` like for previous data) you would have to register it somewhere times when this data is available and iterate to list the available data in the search right?

Reply:
yes, because else we won't be able to know the directory names where our scraper would search data for.

Reply:
Conceptually I don't see why the scraper shouldn't be able to handle a pattern like that, list all directories, use the values. Bonus points for checking that the format is something like `{:d2}` or whatever.